rfcs/0009_Unified_Static_Memory_Planning.md

```diff
@@ -0,0 +1,473 @@
+
+    typedef struct {
+       uint8_t* itcm;
+       uint8_t* flash;
+    }  TVMParameters_my_model;
+
+    typedef struct {
+       uint8_t* input0;
+    }  TVMInputs_my_model;
+
+    typedef struct {
+       uint8_t* output0;
+    }  TVMOutputs_my_model;
+```
+### User Application
+```
+    // The User Application 
+        extern  const TVMModel my_model;
+        __attribute__((section( "ITCM" )  const uint8_t   my_model_params_1[TVM_MY_MODEL_ITCM_PARAMETER_BUFFER_SIZE] = <param_1_data>;
+        __attribute__((section( "FLASH" ), aligned( 16 )))  const uint8_t my_model_params_2[TVM_MY_MODEL_FLASH_PARAMETER_BUFFER_SIZE] = <param_2_data>;
+        __attribute__((section( "DTCM" )  static uint8_t workspace_buffer_1[TVM_MY_MODEL_DTCM_WORKSPACE_BUFFER_SIZE];
+        __attribute__((section( "SRAM" ), aligned( 16 )))  static uint8_t workspace_buffer_2[TVM_MY_MODEL_SRAM_WORKSPACE_BUFFER_SIZE];
+
+    int main(...) {
+         ...
+         TVMContext context;
+         TVMInputs_my_model_1 inputs = {input};
+         TVMOutputs_my_model_1 outputs = {output};
+         TVMWorkspaces_my_model workspaces = {
+             .sram = &workspace_buffer_1,
+             .dtcm = &workspace_buffer_2,
+         };
+         TVMParameters_my_model parameters = {
+             .flash = &my_model_params_1,
+             .itcm = &my_model_params_2
+         };
+         TVMSetWorkspaces(&context, &workspaces);
+         TVMSetParameters(&context, parameters);
+         TVMExecute(&my_model, &inputs, &outputs, &context);
+    }
+```
+# Reference-level explanation
+
+## Overview
+
+This should be a IRModule (TIR) → IRModule (TIR) pass.
+
+Inputs : 
+* AoT TIR PrimFunc ( the control function describing the call graph to operators)
+* All Operator Functions
+* the maximum size for each pool We could use "pinned_memory" (see below) to tag buffers with suggested priority order determined by the scheduler.
+
+The idea is USMP will try to pool them using the preferred "pinned_memory" and fallback whenever the size is exceeding the user provided max size for each pool (if any)
+
+Outputs : 
+* AoT TIR PrimFunc accepting pool buffers from the user.
+* All Operator functions accepting pool buffers.
+  * Each operator function should address using the correct offset in the correct pool buffer
+
+Special Parametric Inputs : 
+* function : The algorithm to be used for planning From a component PoV, the algorithm is a special input with a defined interface.
+
+The current proposal for the interface of the memory planning algorithm is as follows :
+```
+    struct BufferInfo {
+        String name_hint; // this is the tir.buffer name
+        Integer size_bytes;
+        Integer alignment;
+        Array<BufferInfo> conflicts; //the conflicting bufferinfo objs
+        Array<Integer> pool_candidates;`
+        String pool_name;`
+        Integer pool_offset;`
+    }
+```
+```
+Array<BufferInfo> (*foo)(Array<BufferInfo> buffers, Map<String, Integer> pool_sizes)
```

Comment:
One of the assumption is that once the USMP, the scheduling is already decided for the operators.

However, we do understand the requirement.
To that end, this is something we refer to as memory-aware operator scheduling that should be done prior to USMP.

In most cases, this could be done using 'memory pressure' and using let-bindings in relay as relay pass to putting them to A-normal form that results in minimum 'memory pressure'. For advanced (or in other words compile-time tolerable situations), that relay pass could call into the above memory planning interface to retrieve actual allocations.

The reasoning behind that being mixing scheduling and allocations can easily be in-tractable. However, for exceptional cases schedulers are free to be implemented (compile-time in-sensitive) to determine the order by running complex allocation algorithm in-loop. Therefore, we feel this interface is general enough to support both use cases.

Reply:
we discussed this at the µTVM meetup. The current thinking is what @manupa-arm said above--scheduling operators is considered to be separate from memory planning. However, it is entirely reasonable for an upstream scheduler to invoke parts of the memory planner to determine whether reordering the schedule could be advantageous later on. @jroesch also has plans to enable an upstream pass to run an initial "return value" through a partial set of the downstream passes and examine the result of this. For instance, an upstream scheduler could produce an initial schedule and then invoke the remainder of the passes through Memory Planning. Then, it could examine the result of Memory Planning, potentially do this a few times, and choose the best solution as the true "return value" from scheduling.

@MichaelJKlaiberBosch let us know if this seems like a poor compromise to you guys. We certainly don't have to stick with this approach if there are more advantages to another.

Reply:
I am working on a set of RFCs which distill all the pain I felt removing the compile engine. I think users should choose ordering before we plan, entangling passes with different concerns creates some of the problems which we are feeling today, for example the current static memory planner invokes device planning directly and there is no (currently) easy way to split them into pieces. 

If you want to adjust the scheduling you should write a pass to plan, then analyze the plan, rewrite the program and invoke planning again instead of baking these details into the planner. The one shot mentality has led to a lot of super fragile code that only works under a single phase ordering for a single executor.

Reply:
Thanks for the detailed answers.
@areusch @manupa-arm
 In general we are okay with separation scheduling an memory planning problem. From our point of view there could be cases where the memory planner can change the order of the callnodes is of advantage. This, however, depends a lot on the topology of future neural networks. For simple feed-forward NNs this might not be required. More complicated NNs, e.g. as a result of NAS this might be more relevant.